pyro/infer/mcmc/util.py

```diff
@@ -208,3 +212,135 @@ def log_prob(self, model_trace):
         log_probs = self._get_log_factors(model_trace)
         with shared_intermediates() as cache:
             return contract_to_tensor(log_probs, self._enum_dims, cache=cache)
+
+
+def _guess_max_plate_nesting(model, args, kwargs):
+    """
+    Guesses max_plate_nesting by running the model once
+    without enumeration. This optimistically assumes static model
+    structure.
+    """
+    with poutine.block():
+        model_trace = poutine.trace(model).get_trace(*args, **kwargs)
+    sites = [site for site in model_trace.nodes.values()
+             if site["type"] == "sample"]
+
+    dims = [frame.dim
+            for site in sites
+            for frame in site["cond_indep_stack"]
+            if frame.vectorized]
+    max_plate_nesting = -min(dims) if dims else 0
+    return max_plate_nesting
+
+
+def _transform_fn(transforms, params, invert=True):
+    return {k: transforms[k](v) if not invert else transforms[k].inv(v)
+            for k, v in params.items()}
+
+
+def _pe_maker(model, model_args, model_kwargs, trace_prob_evaluator, transforms):
+    def potential_energy(params):
+        params_constrained = {k: transforms[k].inv(v) for k, v in params.items()}
+        cond_model = poutine.condition(model, params_constrained)
```

Comment:
Just a quick note here - `poutine.condition` will convert all stochastic sites into observed sites. I don't think that should really change anything at this point, but just in case you see any weird results, this is worth checking.